### PR TITLE
addressing issue #39427 adding a flag --output to 'kubectl version'

### LINKED ISSUE #39427: Add a short and a json output option for kubectl version
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.):
no

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):
"kubectl version", "kubectl version" json

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): Feature request

**Kubernetes version** (use `kubectl version`):
$ kubectl version --client
Client Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.6", GitCommit:"e569a27d02001e343cb68086bc06d47804f62af6", GitTreeState:"clean", BuildDate:"2016-11-12T05:22:15Z", GoVersion:"go1.7.1", Compiler:"gc", Platform:"darwin/amd64"}

**Environment**:
- **Cloud provider or hardware configuration**:
GKE
- **OS** (e.g. from /etc/os-release):
MacOS 10.12.2

I find it difficult to parse the output of `kubectl version` programmatically. I have to use a YAML parser to do it. I'd prefer to use JSON, or to be given just the string `v1.4.6` and asked to parse it with a regex.

This is obviously a relatively low-priority request.

### PR DESCRIPTION
**Pull request by alejandroEsc**
**What this PR does / why we need it**:
Addressing Issue https://github.com/kubernetes/kubernetes/issues/39427 we all

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #39427 

**Release note**:
```
kubectl version has new flag --output (=json or yaml) allowing result of the command to be parsed in either json format or yaml. 
```

### CODE REVIEW

**Comment by alejandroEsc** [comment]
@k8s-bot bazel test this

**Comment by alejandroEsc** [comment]
There seems to be some issue upstream, never had these tests fail, local shows things working properly.

**Comment by alejandroEsc** [comment]
FYI work in progress for a bit.

**Comment by alejandroEsc** [comment]
@kubernetes/api-reviewers could i get some available eyes here?

File: pkg/version/version.go
Comment (liggitt) [inline pkg/version/version.go]: don't change this, this is being returned from the `/version` API endpoint

File: pkg/kubectl/cmd/version.go
Comment (liggitt) [inline pkg/kubectl/cmd/version.go]: english language keys is odd here... I'd expect something like `client` and `server`

File: pkg/kubectl/cmd/version.go
Comment (liggitt) [inline pkg/kubectl/cmd/version.go]: leave default empty for stdout?

File: pkg/kubectl/cmd/version.go
Comment (liggitt) [inline pkg/kubectl/cmd/version.go]: really? omitempty doesn't work with an empty server struct?

File: pkg/kubectl/cmd/version.go
Comment (liggitt) [inline pkg/kubectl/cmd/version.go]: this doesn't seem right

File: pkg/kubectl/cmd/version.go
Comment (alejandroEsc) [inline pkg/kubectl/cmd/version.go]: it doesnt. You can try it out, i wasnt happy with that either.

Comment (liggitt) [inline pkg/kubectl/cmd/version.go]: you can make it a pointer to the struct and set it to nil when unused, then

File: pkg/kubectl/cmd/version.go
Comment (alejandroEsc) [inline pkg/kubectl/cmd/version.go]: I totally agree, was trying to keep it inline with the original keys. I dont want too much variation between what was expected before and what they will get now. What do you think?

File: pkg/version/version.go
Comment (alejandroEsc) [inline pkg/version/version.go]: you mean all these vars? do I make the changes there then?

Comment (liggitt) [inline pkg/version/version.go]: no, revert this change. it's ok to print empty strings for these fields

File: pkg/kubectl/cmd/version.go
Comment (alejandroEsc) [inline pkg/kubectl/cmd/version.go]: :face_palm: arggghhh sorry about this.

File: pkg/kubectl/cmd/version.go
Comment (alejandroEsc) [inline pkg/kubectl/cmd/version.go]: i was going back and forth on this. I settled on std simply because i thought empty string would be a bit more difficult to understand? Not sure, but i'll change it to empty string if you are ok with that?

**Comment by alejandroEsc** [comment]
@liggitt  left some questions there for you just to get some direction, thank you soooo much for taking your time to review, much appreciated!

**Comment by alejandroEsc** [comment]
@liggitt one more question, have you taken a look at the issue ticket, just wanted your opinion on the output now vs older format.

File: pkg/kubectl/cmd/version.go
Comment (alejandroEsc) [inline pkg/kubectl/cmd/version.go]: @liggitt ahhh, good point, will try this.

File: pkg/kubectl/cmd/version.go
Comment (liggitt) [inline pkg/kubectl/cmd/version.go]: for programmatic output, I would go short and lowercase

**Comment by alejandroEsc** [comment]
@liggitt made changes requested, please have another look, and thank you again!

File: pkg/kubectl/cmd/version.go
Comment (liggitt) [inline pkg/kubectl/cmd/version.go]: I would order it like this:
1. get the client and (optionally) server versions
2. build the VersionObj struct with nil set appropriately
3. switch on cmdutil.GetFlagString(cmd, "output") (don't normalize their arg)
4. in the "" case, have all the existing output, including short output handling
5. in the json/yaml cases, marshal and output
6. in the default case, error

Comment (alejandroEsc) [inline pkg/kubectl/cmd/version.go]: That would be the ideal case. However, the "" case makes it a bit difficult because the output is actually `label: version.Info{...}` which makes this approach difficult to do. I have to keep the old output because legacy tests etc.. still use this.

Comment (liggitt) [inline pkg/kubectl/cmd/version.go]: Right, in the "" case, include all the old output code

Comment (alejandroEsc) [inline pkg/kubectl/cmd/version.go]: Ok, i think i understand what you are saying... let me try it out. it will most likely be submitted later tonight though, fyi.

File: pkg/kubectl/cmd/version.go
Comment (k82cn) [inline pkg/kubectl/cmd/version.go]: Please avoid duplicated code of generating version info.

**Comment by alejandroEsc** [comment]
ok @liggitt that was fun, take a look when you are ready and thank you again.

File: pkg/kubectl/cmd/version.go
Comment (liggitt) [inline pkg/kubectl/cmd/version.go]: Only handle short output in the human readable case

File: pkg/kubectl/cmd/version.go
Comment (liggitt) [inline pkg/kubectl/cmd/version.go]: Store this, but don't return it yet. Otherwise a server error prevents us from seeing our client version and is a behavior change

File: pkg/kubectl/cmd/version.go
Comment (liggitt) [inline pkg/kubectl/cmd/version.go]: Same, don't truncate data here. Truncate in short mode only in the human readable case

File: pkg/kubectl/cmd/version.go
Comment (liggitt) [inline pkg/kubectl/cmd/version.go]: Can just check whether you have a non nil server version

File: pkg/kubectl/cmd/version.go
Comment (liggitt) [inline pkg/kubectl/cmd/version.go]: Return server error here if you got one

Comment (liggitt) [inline pkg/kubectl/cmd/version.go]: Actually, return server error after the switch statement

**Review by liggitt — CHANGES_REQUESTED** [review]

File: pkg/kubectl/cmd/version.go
Comment (alejandroEsc) [inline pkg/kubectl/cmd/version.go]: Can you clarify why this is the case?

Comment (liggitt) [inline pkg/kubectl/cmd/version.go]: The only point of json/yaml output is to consume specific fields programmatically. Truncating to a single field doesn't make that any easier

Comment (alejandroEsc) [inline pkg/kubectl/cmd/version.go]: I disagree, regardless of the consumer we should respect the expectations of the flag.

**Comment by alejandroEsc** [comment]
Very good points, I'll try to make changes before the end of the day.

File: pkg/kubectl/cmd/version.go
Comment (liggitt) [inline pkg/kubectl/cmd/version.go]: Then disallow short output combined with json/yaml output

**Comment by alejandroEsc** [comment]
@liggitt made the changes you suggested and ready for your review.

File: pkg/kubectl/cmd/version.go
Comment (liggitt) [inline pkg/kubectl/cmd/version.go]: retrieve the server version once at the top of the function, just save the serverErr and don't return early. the function should look like this:

1. get client version
2. if needed, get server version
3. switch on output type and output
4. return serverErr if we got an error

Comment (alejandroEsc) [inline pkg/kubectl/cmd/version.go]: good point.

**Comment by alejandroEsc** [comment]
@liggitt changes made.

File: pkg/kubectl/cmd/version.go
Comment (liggitt) [inline pkg/kubectl/cmd/version.go]: put the fetch inside the `if` block... we don't want to hit the server if they've told us they only want the client version

File: pkg/kubectl/cmd/version.go
Comment (liggitt) [inline pkg/kubectl/cmd/version.go]: no need to check the flags or errors here, just output the serverVersion if we have a non-nil one

File: pkg/kubectl/cmd/version.go
Comment (liggitt) [inline pkg/kubectl/cmd/version.go]: move this after the switch, since all cases share it

Comment (alejandroEsc) [inline pkg/kubectl/cmd/version.go]: the human readable output requires different logic, so not all switch cases share this.

File: pkg/kubectl/cmd/version.go
Comment (alejandroEsc) [inline pkg/kubectl/cmd/version.go]: yep, true.

File: pkg/kubectl/cmd/version.go
Comment (alejandroEsc) [inline pkg/kubectl/cmd/version.go]: original logic displayed an error at this point and returned that error. I want to keep original behavior here (addresses one of your comments below).

**Comment by alejandroEsc** [comment]
@liggitt ok, reviewed and made changes, please take a look at my comments regarding moving the serverErr out of the code block.

**Comment by alejandroEsc** [comment]
@liggitt changes committed as we discussed, thanks!

**Comment by liggitt** [comment]
change looks good, go ahead and squash commits, and add a test to test-cmd-util.sh exercising the new options (add a new test function sort of like `run_kubectl_run_tests`, named something like `run_kubectl_version_tests`, and exercise the short, long, client-only, json, and yaml output

**Comment by alejandroEsc** [comment]
awesome, thank you for pointing me to the test!

**Comment by bgrant0607** [comment]
cc @pwittrock @fabianofranz @AdoHe

**Comment by AdoHe** [comment]
generally looks good to me, @alejandroEsc could you please add some tests?

**Comment by alejandroEsc** [comment]
@AdoHe working on it as we speak based on discussions with @liggitt. Would you be ok to check the tests?

**Comment by alejandroEsc** [comment]
@liggitt @AdoHe I am not bash testing expert. So this is my first real attempt at this. Could you please review and let me know of any concern. Additionally if it is not the direction you think this should go would you mind taking this discussion on slack?

**Comment by alejandroEsc** [comment]
@k8s-bot gci gce e2e test this

**Comment by ixdy** [comment]
I don't have strong opinions on this, but from my PR where I was working on something similar (#37049), there were some suggestions by @fabianofranz, @juanvallejo, and @philips to possibly use the existing printer (pkg/kubectl/resource_printer.go) to print the version. Doing such would require inventing a version schema object, which may be more work than it's worth.

There was also a suggestion to deprecate the `--short` flag and fold it under `--output`.

**Comment by alejandroEsc** [comment]
@fabianofranz @juanvallejo @philips please take a look, any additional input welcomed. For additional context, `--short` is still available, but will not make changes if the `--output` flag has been given also. Currently I am waiting for reviewers to look at tests which is the last ask for this PR.

**Comment by alejandroEsc** [comment]
@liggitt @AdoHe ready for review when you have a chance, thank you.

File: hack/lib/test.sh
Comment (liggitt) [inline hack/lib/test.sh]: this file is for generic test helper funcs, not specific tests

File: hack/lib/test.sh
Comment (liggitt) [inline hack/lib/test.sh]: this is overly fragile, what are you trying to test?

File: hack/lib/test.sh
Comment (liggitt) [inline hack/lib/test.sh]: sorting internally is unexpected

File: hack/make-rules/test-cmd-util.sh
Comment (liggitt) [inline hack/make-rules/test-cmd-util.sh]: testing the output of multiple invocations of version against itself is a little weird

in general, it would be easier to follow to do something like 
```
kubectl "${kube_flags[@]}" version ... > version.output
fgrep "expected output" version.output 
```

Comment (alejandroEsc) [inline hack/make-rules/test-cmd-util.sh]: expected output would change for each version, including testing on different machines. Any ideas on how to handle this would be awesome?

File: hack/lib/test.sh
Comment (alejandroEsc) [inline hack/lib/test.sh]: I dont expect that output to always be ordered, so i have to get it sort it so i can get some way to compare the data.

File: hack/lib/test.sh
Comment (alejandroEsc) [inline hack/lib/test.sh]: the function isnt a test, its a helper function. I suppose i can rename it to make that clear.

File: hack/lib/test.sh
Comment (alejandroEsc) [inline hack/lib/test.sh]: It isnt a test. Its trying to get either the client or server version of output, which should be stable since we have tests that depend on this output being correct.

**Comment by liggitt** [comment]
cc @kubernetes/sig-cli-api-reviews

**Comment by liggitt** [comment]
the shape of this seems reasonable to me, tagging in sig-cli folks for approval

**Comment by alejandroEsc** [comment]
@fabianofranz @AdoHe pinging for approval

File: pkg/kubectl/cmd/version.go
Comment (juanvallejo) [inline pkg/kubectl/cmd/version.go]: Maybe check outside of the switch if both `--short` and `--output` were specified?
An error could be returned if both `--short` and `--output` are specified

File: pkg/kubectl/cmd/version.go
Comment (juanvallejo) [inline pkg/kubectl/cmd/version.go]: This might benefit from being in a goroutine

File: pkg/kubectl/cmd/version.go
Comment (philips) [inline pkg/kubectl/cmd/version.go]: Please just call it Version. I haven't seen Obj as a suffix in the rest of the codebase.

**Comment by philips** [comment]
My only wondering is if the `Version struct` should go into the apimachinery Version package. Then someone would have all structs to parse this in that one pkg.

**Comment by AdoHe** [comment]
```
My only wondering is if the Version struct should go into the apimachinery Version package. Then someone would have all structs to parse this in that one pkg.
```
@philips yep, this is just something we should consider.

File: pkg/kubectl/cmd/version.go
Comment (AdoHe) [inline pkg/kubectl/cmd/version.go]: SGTM.

File: pkg/kubectl/cmd/version.go
Comment (alejandroEsc) [inline pkg/kubectl/cmd/version.go]: I dont think there would be much gained by adding an error if both flags are used. The suggestion goes back to my disagreement regarding expected/unexpected behavior using flags i had prior, please look at the history, regarding how it is implemented now. If you feel strongly about this, I think it would be best we get together and discuss it in detail and perhaps discuss some enforcement throughout the code, or what exists atm.

File: pkg/kubectl/cmd/version.go
Comment (alejandroEsc) [inline pkg/kubectl/cmd/version.go]: agreed.

**Comment by alejandroEsc** [comment]
@AdoHe @philips to clarify is that last suggestion for the future, or an item you want me to approach for now?

File: pkg/kubectl/cmd/version.go
Comment (alejandroEsc) [inline pkg/kubectl/cmd/version.go]: very interesting idea, I would say to keep things clean and within scope, that we should open a new PR for that if we feel strongly about it.

**Comment by alejandroEsc** [comment]
@k8s-bot unit test this

File: pkg/kubectl/cmd/version.go
Comment (juanvallejo) [inline pkg/kubectl/cmd/version.go]: @alejandroEsc 
> I would say to keep things clean and within scope, that we should open a new PR for that if we feel strongly about it.

Leaving this for a separate PR sounds good to me

**Comment by alejandroEsc** [comment]
@k8s-bot unit test this

**Comment by alejandroEsc** [comment]
@k8s-bot cvm gce e2e test this

**Comment by alejandroEsc** [comment]
@AdoHe @philips i made the changes requested and rebased along with the newest changes to version.go. I would like to know if it would be possible to just write go tests instead of bash ones? What do you guys think?

**Comment by alejandroEsc** [comment]
@k8s-bot gci gke e2e test this

**Comment by alejandroEsc** [comment]
@k8s-bot cross build this

**Comment by alejandroEsc** [comment]
@liggitt @AdoHe @philips could i get a /lgtm on this?

**Comment by alejandroEsc** [comment]
@liggitt suggestions on how to proceed. Do not want to get stuck rebasing.

**Comment by AdoHe** [comment]
@liggitt this generally looks good to me, any more suggestions?

**Comment by liggitt** [comment]
> @liggitt this generally looks good to me, any more suggestion?

Nothing else from me

**Comment by alejandroEsc** [comment]
@AdoHe if its alright with you then, can I get an approval/lgtm from you?

**Comment by alejandroEsc** [comment]
Could someone from @kubernetes/sig-cli-pr-reviews take a quick look and potentially give an lgtm before the deadline tonight? Would be nice to have this available for 1.6

**Comment by AdoHe** [comment]
@alejandroEsc sorry about respond so late, I was on a business trip, not have too much time to check email. :(

**Comment by alejandroEsc** [comment]
@k8s-bot verify test this

**Comment by alejandroEsc** [comment]
@AdoHe no problem! i figured you were out, it happens, just wanted to meet cut-off for this release but looks like that wont happen, oh well.

**Comment by alejandroEsc** [comment]
@AdoHe could you give the PR and approval?

**Comment by alejandroEsc** [comment]
@deads2k @lavalamp could i get an approval from one of you.

**Comment by deads2k** [comment]
/approve

But I don't think made freeze.

**Comment by alejandroEsc** [comment]
@deads2k correct, I didnt. However i have requested an exception.

**Comment by ethernetdan** [comment]
A 1.6 code freeze exception has been submitted for this PR - please let @kubernetes/release-team know of any status changes.

**Comment by alejandroEsc** [comment]
@k8s-bot cvm gce e2e test this
@k8s-bot gci gce e2e test this

**Comment by alejandroEsc** [comment]
@k8s-bot cross build this
@k8s-bot gce etcd3 e2e test this

**Comment by alejandroEsc** [comment]
@adohe could I get the lgtm applied again? Sorry for the trouble.

**Comment by alejandroEsc** [comment]
@ethernetdan should you remove that donotmerge label and add a 1.7 label instead? also the bot removed the lgtm because of a flake.

File: pkg/kubectl/cmd/version.go
Comment (alejandroEsc) [inline pkg/kubectl/cmd/version.go]: @eparis yeah probably got deleted and re-added after some change, ill go ahead and make the format change.

File: pkg/kubectl/cmd/version.go
Comment (eparis) [inline pkg/kubectl/cmd/version.go]: `encoding/json` is part of the standard lib.

Comment (alejandroEsc) [inline pkg/kubectl/cmd/version.go]: ha, i totally missed that during moving things around. thanks for the catch. Could you give this pr and lgtm? i got this dropped after an automatic retest.

**Comment by kargakis** [comment]
@alejandroEsc does this also address https://github.com/kubernetes/kubernetes/issues/43750? Can you post the output of `kubectl version` with your fix?

**Comment by kargakis** [comment]
\lgtm

**Comment by alejandroEsc** [comment]
@kargakis please look at the original ticket: https://github.com/kubernetes/kubernetes/issues/39427, here i post the output. Thanks again for having a look, hopefully we arent hit with a flake again.

**Comment by kargakis** [comment]
Those are looking good but I am asking specifically about the output of the command w/o using -o (the default case)

**Comment by alejandroEsc** [comment]
the default, without any additional output has not changed, the intend here is to keep default behavior and only allow the option to make it human readable if we want it.

**Comment by kargakis** [comment]
That's what I thought. Thanks for clarifying. cc: @php-coder

**Comment by alejandroEsc** [comment]
@k8s-bot bazel test this

**Comment by alejandroEsc** [comment]
@k8s-bot bazel test this

**Comment by alejandroEsc** [comment]
@k8s-bot bazel test this

what is going on with bazel?

**Comment by ixdy** [comment]
Looks like the bazel build is just broken - the upstream busybox dependency no longer exists. cc @mikedanese @spxtr 
```
W0328 18:37:19.565] ERROR: /workspace/k8s.io/kubernetes/build/BUILD:21:1: no such package '@busybox_deb//file': Error downloading [http://ftp.us.debian.org/debian/pool/main/b/busybox/busybox-static_1.22.0-19+b1_amd64.deb] to /root/.cache/bazel/_bazel_root/e9f728bbd90b3fba632eb31b20e1dacd/external/busybox_deb/busybox-static_1.22.0-19+b1_amd64.deb: GET returned 404 Not Found and referenced by '//build:busybox'.
W0328 18:37:19.875] ERROR: Analysis of target '//build/release-tars:release-tars' failed; build aborted.
```

**Pull request merged**
